### PR TITLE
Prototype React version of tronweb

### PR DESCRIPTION
**Pull request by qui**
I implemented a basic version of the homepage and part of a Job view, to see how feasible it would be to re-write Tronweb in React.

Will definitely take some time to recreate all the interactive components, but I think we can simplify the pages as we do this.

I'm running this version at http://tronplayground-uswest1adevc.dev.yelpcorp.com:8088/web2/.

### CODE REVIEW

File: yelp_package/bionic/Dockerfile [hunk 1]
```diff
@@ -27,5 +27,12 @@ RUN apt-get -q update && \
 ARG PIP_INDEX_URL
 ENV PIP_INDEX_URL=${PIP_INDEX_URL:-https://pypi.python.org/simple}
 
+# Get yarn and node
+RUN wget --quiet -O - https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+RUN wget --quiet -O - https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
+RUN echo "deb http://dl.yarnpkg.com/debian/ stable main" > /etc/apt/sources.list.d/yarn.list
+RUN echo "deb http://deb.nodesource.com/node_8.x bionic main" > /etc/apt/sources.list.d/nodesource.list
+RUN apt-get -q update && apt-get -q install -y --no-install-recommends yarn nodejs
```
[the remark this answers is not in the document]
Comment (qui) [reply yelp_package/bionic/Dockerfile:35]: No, adding yarn/node in the initial list at the top doesn't work. I got these sources from the YARN docs about installation: https://classic.yarnpkg.com/en/docs/install/#debian-stable

File: tronweb2/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    "react-dom": "^16.13.1",
+    "react-router-dom": "^5.2.0",
+    "react-scripts": "3.4.1",
+    "jquery": "3.5.1",
+    "d3": "^5.16.0"
+  },
+  "scripts": {
+    "start": "react-scripts start",
+    "build": "react-scripts build",
+    "test": "react-scripts test",
```
[the remark this answers is not in the document]
Comment (qui) [reply tronweb2/package.json:17]: No on the tests, there are no tests right now for the js.

Just updated the review to run the eslint in pre-commit, which is part of the tox test env.

File: tronweb2/src/components/ActionGraph/ActionGraph.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+function setNodeLevel(action, actionMap) {
+    if ("level" in action) {
+        return action.level;
+    }
+    if (action.dependencies.length === 0) {
+        action.level = 0;
+        return 0;
+    }
+    var dependencyLevels = action.dependencies.map((actionName) => {
```
[the remark this answers is not in the document]
Comment (qui) [reply tronweb2/src/components/ActionGraph/ActionGraph.js:14]: Letting the linters do this is so much nicer, thank you!

File: tronweb2/src/components/JobsDashboard/JobsDashboard.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import {getJobColor} from '../../utils/utils';
+import JobScheduler from '../JobScheduler';
+
+function JobsDashboard() {
+  const [jobData, setJobData] = useState(undefined);
+
+  useEffect(() => {
+    fetch('/api/jobs')
+      .then(response => response.json())
+      .then(data => setJobData(data));
```
[the remark this answers is not in the document]
Comment (qui) [reply tronweb2/src/components/JobsDashboard/JobsDashboard.js]: Thanks, updated it to handle those cases (after moving into a util function). I tried async/await, but I find it a little bit easier to reason about how the UI will update this way, so I'm going to leave the chaining.

**Comment by askedrelic** [comment]
I'll +1, getting a stable rendering base for React apps can be a challenge, so this is good. There is also design and splitting of components, which is also good. A next possible step iscreating more "data model" components to handle fetching of data, separating data fetching from rendering, but fetching inline is fine for now.

File: tronweb2/src/components/Job/Job.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  );
+}
+
+function Job() {
+  const [job, setJobData] = useState(undefined);
+  const { jobId } = useParams();
+
+  useEffect(() => {
+    document.title = jobId;
+    fetchFromApi(`/api/jobs/${jobId}?include_action_graph=1`, setJobData);
```
Comment (mxmul) [inline tronweb2/src/components/Job/Job.js:70]: nit: it's probably not a real issue, but there is a race condition here: if the `Job` component is unmounted before this `fetch` resolves, the `setJobData` call will fail. To be totally safe, you'd want to [cancel the `fetchFromApi` promise](https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html) in this effects [cleanup](https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup).

Comment (qui) [reply tronweb2/src/components/Job/Job.js]: 👍 The Tron API can be pretty slow at times, so might as well handle this case - updated. Tested by making the tron playground master sleep before responding to requests, and got a warning before and no warning after the change, when I clicked away from the job page before it loaded.

**Pull request merged**